camel-core/src/main/java/org/apache/camel/model/transformer/CustomTransformerDefinition.java

```diff
@@ -0,0 +1,106 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.model.transformer;
+
+import java.util.Map;
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlAnyAttribute;
+import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlTransient;
+import javax.xml.bind.annotation.XmlType;
+import javax.xml.namespace.QName;
+
+import org.apache.camel.CamelContext;
+import org.apache.camel.Endpoint;
+import org.apache.camel.component.bean.BeanHolder;
+import org.apache.camel.component.bean.BeanProcessor;
+import org.apache.camel.component.bean.ConstantStaticTypeBeanHolder;
+import org.apache.camel.component.bean.ConstantTypeBeanHolder;
+import org.apache.camel.impl.transformer.ProcessorTransformer;
+import org.apache.camel.spi.DataFormat;
+import org.apache.camel.spi.DataType;
+import org.apache.camel.spi.Metadata;
+import org.apache.camel.spi.RouteContext;
+import org.apache.camel.spi.Transformer;
+import org.apache.camel.util.ObjectHelper;
+
+/**
+ * Represents a BeanTransformer.
+ */
+@Metadata(label = "transformation")
+@XmlType(name = "customTransformer")
+@XmlAccessorType(XmlAccessType.FIELD)
+public class CustomTransformerDefinition extends TransformerDefinition {
+    @XmlAttribute
+    private String ref;
+    @XmlAttribute
+    private String type;
+
+    @Override
+    protected Transformer doCreateTransformer() throws Exception {
+        if (ref == null && type == null) {
+            throw new IllegalArgumentException("'ref' or 'type' must be specified for customTransformer");
+        }
+        Transformer transformer;
+        if (ref != null) {
+            transformer = getCamelContext().getRegistry().lookupByNameAndType(ref, Transformer.class);
+            if (transformer == null) {
+                throw new IllegalArgumentException("Cannot find transformer with ref:" + ref);
+            }
+            if (transformer.getModel() != null || transformer.getFrom() != null || transformer.getTo() != null) {
+                throw new IllegalArgumentException(String.format("Transformer '%s' is already in use. Please check if duplicate transformer exists.", ref));
+            }
+        } else {
+            Class<Transformer> transformerClass = getCamelContext().getClassResolver().resolveMandatoryClass(type, Transformer.class);
+            if (transformerClass == null) {
+                throw new IllegalArgumentException("Cannot find transformer class: " + type);
+            }
+            transformer = transformerClass.newInstance();
```

Comment:
Use camelContext injector to create a new instance as it can integrates with spring/blueprint etc and supports IoC stuff